### PR TITLE
feat(api): new endpoint PartialUpdateHost

### PR DESCRIPTION
**Pull request by a-launois**
## Description

New configuration endpoint PartialUpdateHost
```
PATCH /configuration/hosts/{host_id}
```

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [x] 24.04.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php [hunk 1]
```diff
@@ -0,0 +1,616 @@
...
+                $macro->setDescription(
+                    isset($commandMacros[$macro->getName()])
+                    ? $commandMacros[$macro->getName()]->getDescription()
+                    : ''
+                );
+                // TODO : see which is better for phpstan / cs fixer
+                // $commandMacro = $commandMacros[$macro->getName()] ?? null;
+                // $macro->setDescription(
+                //     $commandMacro ? $commandMacro->getDescription() : ''
+                // );
```
Comment (a-launois) [inline centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php:569]:
```suggestion
```

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php [hunk 2]
```diff
@@ -0,0 +1,616 @@
...
+
+            $this->writeMonitoringServerRepository->notifyConfigurationChange($host->getMonitoringServerId());
+            if ($previousMonitoringServer !== $host->getMonitoringServerId()) {
+                // Monitoring server has change, notify previous monitoring server of configuration changes.
+                 $this->writeMonitoringServerRepository->notifyConfigurationChange($previousMonitoringServer);
+            }
+
+            $this->dataStorageEngine->commitTransaction();
+        } catch (\Throwable $ex) {
+            $this->error("Rollback of 'PartialUpdateHost' transaction.");
```
Comment (callapa) [inline centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php:177]:
```suggestion
            $this->error("Rollback of 'PartialUpdateHost' transaction.", ['trace' => $ex->getTraceAsString()]);
```

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php [hunk 3]
```diff
@@ -0,0 +1,611 @@
...
+     * @throws \Throwable|AssertionFailedException|HostException
+     */
+    private function updateHost(PartialUpdateHostRequest $dto, Host $host): void
+    {
+        $this->info('PartialUpdateHost: update host', ['host_id' => $host->getId()]);
+
+        $inheritanceMode = $this->optionService->findSelectedOptions(['inheritance_mode']);
+        $inheritanceMode = isset($inheritanceMode[0])
+            ? (int) $inheritanceMode[0]->getValue()
+            : 0;
```
Comment (callapa) [inline centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php:197]: What does the value 0 mean for inheritance mode?

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php [hunk 4]
```diff
@@ -0,0 +1,611 @@
...
+            $host->setNotificationOptions(
+                $dto->notificationOptions === null
+                    ? []
+                    : HostEventConverter::fromBitFlag($dto->notificationOptions)
+            );
+        }
+
+        if (! $dto->addInheritedContactGroup instanceOf NoValue) {
+            $host->setAddInheritedContactGroup(
+                $inheritanceMode === 1 ? $dto->addInheritedContactGroup : false
```
Comment (callapa) [inline centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php:377]: What does the value 1 mean for inheritance mode?

File: centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostRepository.php [hunk 5]
```diff
@@ -115,6 +133,58 @@ public function deleteById(int $hostId): void
...
+            $this->updateBasicInformations($host);
+            $this->updateExtendedInformations($host);
+            $this->updateMonitoringServer($host);
+            $this->updateSeverity($host);
+
+            if (! $alreadyInTransaction) {
+                $this->db->commit();
+            }
+        } catch (\Throwable $ex) {
+             $this->error($ex->getMessage(), ['trace' => $ex->getTraceAsString()]);
```
Comment (callapa) [inline centreon/src/Core/Host/Infrastructure/Repository/DbWriteHostRepository.php:156]: Log redundant with try catch in use case

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php (hunk 4 above)
Comment (a-launois) [reply centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php:377]: 1 = Vertical Inheritance Only
2 = Closest Value
3 = Cumulative inheritance

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php (hunk 3 above)
Comment (a-launois) [reply centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php:197]: I'll change it to null

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php (hunk 4 above)
Comment (a-launois) [reply centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php]: Changed it for a const

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/HostMacroFactory.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace Core\Host\Application\UseCase\PartialUpdateHost;
+
+use Assert\AssertionFailedException;
+use Core\Macro\Domain\Model\Macro;
+
+final class HostMacroFactory
```
Comment (arnapou) [inline centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/HostMacroFactory.php:29]: I just noticed taht we know have 4 classes named `HostMacroFactory`.
I don't think that's a good idea, but anyway, forget it and just resolve this comment.
I couldn't make it as if I saw nothing, not your fault.

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php [hunk 7]
```diff
@@ -0,0 +1,617 @@
...
+            $this->updateHost($request, $host);
+            $this->updateHostCategories($request, $host);
+            $this->updateHostGroups($request, $host);
+            $this->updateParentTemplates($request, $host);
+            // Note: parent templates must be updated before macros for macro inheritance resolution
+            $this->updateMacros($request, $host);
+
+            $this->writeMonitoringServerRepository->notifyConfigurationChange($host->getMonitoringServerId());
+            if ($previousMonitoringServer !== $host->getMonitoringServerId()) {
+                // Monitoring server has change, notify previous monitoring server of configuration changes.
```
Comment (arnapou) [inline centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHost.php:177]: Typo
```suggestion
                // Monitoring server has changed, notify previous monitoring server of configuration changes.
```

File: centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHostValidation.php [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    }
+
+    /**
+     * Assert monitoring server ID is valid.
+     *
+     * @param int $monitoringServerId
+     *
+     * @throws HostException|\Throwable
+     */
+    public function assertIsValidMonitoringServer(int $monitoringServerId): void
```
Comment (arnapou) [inline centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHostValidation.php:113]: Apparent unused method `assertIsValidMonitoringServer`.
Can you check it out and eventualy remove it ?

Comment (a-launois) [reply centreon/src/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHostValidation.php:113]: I forgot to call it 😞

**Review by arnapou — DISMISSED** [review]

File: centreon/tests/php/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHostTest.php [hunk 9]
```diff
@@ -0,0 +1,825 @@
...
+        ->and($this->presenter->response->getMessage())
+        ->toBe(
+            HostException::idsDoNotExist(
+                'templates',
+                $this->request->templates
+            )->getMessage()
+        );
+});
+
+it('should present a ConflictResponse when a parent template create a circular inheritance', function (): void {
```
Comment (cg-tw) [inline centreon/tests/php/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHostTest.php:668]:
```suggestion
it('should present a ConflictResponse when a parent template creates a circular inheritance', function (): void {
```

File: centreon/tests/php/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHostValidationTest.php [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+        ->method('exist')
+        ->willReturn([]);
+
+    $this->validation->assertAreValidTemplates([1,3], 4);
+})->throws(
+    HostException::class,
+    HostException::idsDoNotExist('templates', [1,3])->getMessage()
+);
+
+it('throws an exception when parent template ID create a circular inheritance', function (): void {
```
Comment (cg-tw) [inline centreon/tests/php/Core/Host/Application/UseCase/PartialUpdateHost/PartialUpdateHostValidationTest.php:263]:
```suggestion
it('throws an exception when parent template ID creates a circular inheritance', function (): void {
```

**Review by cg-tw — APPROVED** [review]

**Review by arnapou — APPROVED** [review]

**Review by callapa — APPROVED** [review]

**Pull request merged**